### DISCUSSION TITLE
Should be blank string as enum member be used in AJV schema validation

### DISCUSSION THREAD

**Discussion by din-s** [json-schema-org/community#579] [General]
I have a specific requirement in schema that with respect to 2 properties
schema: {
property1: {type: string, enum: ["val1", "val2", ""]}
property2: {type: string, enum: ["value1", "value2", ""]}
...//other properties
}

I wish to accomplish a state of validation such that only either of property1, property2 exists and at the same when one contains value other must be empty.

But I think keeping a "" as enum is not an ethical check.

**Comment by gregsdennis** [comment]
JSON Schema doesn't provide a way to reference data like this out of the box.  You can do it with my [Data vocab extension](https://docs.json-everything.net/schema/vocabs/data-2023/), though.  AJV has their own (`$data`), but it doesn't create valid schemas.  (I can go into why later, if needed.)  But neither of these are really needed for this particular problem.  Your situation has finite cases, so it can be expanded finitely.

To do this with off-the-shelf JSON Schema, you just need to approach the problem differently.  It can be solved by expanding it out a bit.  The approach you need to take is:

```json
{
  "type": "object",
  "anyOf": [
    {
      "properties": {
        "prop1": { "enum": [ "val1", "val2" ] },
        "prop2": { "const": "" }
      }
    },
    {
      "properties": {
        "prop1": { "const": "" },
        "prop2": { "enum": [ "value1", "value2" ] }
      }
    }
  ]
}
```

As you add more data dependencies, this will grow, but it can usually be expanded like this.

If you find that you have a situation where you need to validate something like "prop1 mod prop2 = 0", then you'll need the data vocab.

**Comment by jdesrosiers** [comment]
Expressing that only one of two properties is present is fairly easy,

```json
{
  "type": "object",
  "properties": {
    "property1": { "enum": ["val1", "va2"],
    "proprety2": { "enum": ["value1", "value2"]
  },

  "oneOf": [
    { "required": ["property1"] },
    { "required": ["property2"] }
  ]
}
```

However, you seem to be using the empty string (`""`) to represent the value as not being present. That makes this more difficult to express. You end up with something similar to what Greg expressed, but I would organize it differently. I like to have all the basic constraints at the top and express complex assertions like this one separately.

```json
{
  "type": "object",
  "properties": {
    "property1": { "enum": ["val1", "va2"] },
    "proprety2": { "enum": ["value1", "value2"] }
  },

  "oneOf": [
    {
      "required": ["property1"],
      "property2": { "const": "" }
    },
    {
      "required": ["property2"],
      "property1": { "const": "" }
    }
  ]
}
```
